src/shims/foreign_items.rs

```diff
@@ -744,6 +744,60 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
                     dest,
                 )?;
             }
+            "memcpy" => {
+                let [ptr_dest, ptr_src, n] =
+                    this.check_shim(abi, Abi::C { unwind: false }, link_name, args)?;
+                let ptr_dest = this.read_pointer(ptr_dest)?;
+                let ptr_src = this.read_pointer(ptr_src)?;
+                let n = this.read_target_usize(n)?;
+                this.mem_copy(
+                    ptr_src,
+                    Align::ONE,
+                    ptr_dest,
+                    Align::ONE,
+                    Size::from_bytes(n),
+                    true,
+                )?;
+                this.write_pointer(ptr_dest, dest)?;
+            }
+            "strcpy" => {
+                let [ptr_dest, ptr_src] =
+                    this.check_shim(abi, Abi::C { unwind: false }, link_name, args)?;
+                let ptr_dest = this.read_pointer(ptr_dest)?;
+                let ptr_src = this.read_pointer(ptr_src)?;
+                let n = this.read_c_str(ptr_src)?.len().saturating_add(1);
+                this.mem_copy(
+                    ptr_src,
+                    Align::ONE,
+                    ptr_dest,
+                    Align::ONE,
+                    Size::from_bytes(n),
+                    true,
+                )?;
```

Comment:
This will preserve provenance in the copy. Is that really what we want?

Might make more sense to to `read_os_str_from_c_str` and then `write_os_str_to_c_str`.

Reply:
It's not true that they only deal with strings. Even if you are thinking based on strict aliasing, these functions deal in `char *` which means they can copy any object.

Whether using these functions like that is a _good idea_ is a separate question.

Reply:
Hm yeah seems okay to make it public.

I guess we can also allow copying pointers with strcpy, if you think that's less confusing :shrug: